src/site/content/en/blog/porting-gphoto2-to-the-web/index.md

```diff
@@ -0,0 +1,352 @@
+
+Only after looking at the source code of the corresponding function in the console utility, I found out that it's not actually getting a video at all, but instead [keeps retrieving camera's preview](https://github.com/gphoto/gphoto2/blob/e9ad2a460990afec3d09033000642297b16d8950/gphoto2/actions.c#L1068-L1070) as individual JPEG images in an endless loop, and writing them out one by one so that they form an [M-JPEG](https://en.wikipedia.org/wiki/Motion_JPEG) stream:
+
+```c
+while (1) {
+  const char *mime;
+  r = gp_camera_capture_preview (p->camera, file, p->context);
+  // …
+```
+
+I was very surprised that this approach works efficiently enough to get an impression of smooth realtime video. I was even more skeptical about being able to match the same performance in the web application too, with all the extra abstractions and Asyncify in the way. However, I decided to try anyway.
+
+On the C++ side I exposed a method `capturePreviewAsBlob` that invokes the same `gp_camera_capture_preview` function, and converts the resulting in-memory file to a [`Blob`](https://developer.mozilla.org/docs/Web/API/Blob) that can be passed to other web APIs more easily:
+
+```cpp
+val capturePreviewAsBlob() {
+  return gpp_rethrow([=]() {
+    auto &file = get_file();
+
+    gpp_try(gp_camera_capture_preview(camera.get(), &file, context.get()));
+
+    auto params = blob_chunks_and_opts(file);
+    return Blob.new_(std::move(params.first), std::move(params.second));
+  });
+}
+```
+
+On the JavaScript side, I have a loop, similar to the one in gPhoto2, that keeps retrieving preview images as `Blob`s, decodes them in the background with [`createImageBitmap`](https://developer.mozilla.org/docs/Web/API/createImageBitmap) and [transfers](https://developer.mozilla.org/docs/Web/API/ImageBitmapRenderingContext/transferFromImageBitmap) them to the canvas on the next animation frame:
+
+```js
+while (this.canvasRef.current) {
+  try {
+    let blob = await this.props.getPreview();
+
+    let img = await createImageBitmap(blob, { /* … */ });
+    await new Promise(resolve => requestAnimationFrame(resolve));
+    canvasCtx.transferFromImageBitmap(img);
+  } catch (err) {
+    // …
+  }
+}
+```
+
+Using those modern APIs ensures that all of the decoding work is done in the background, and the canvas gets updated only when both the image and the browser are fully prepared for drawing. This achieved consistent 30+ FPS on my laptop, which matched the native performance of both gPhoto2 and the official Sony software.
+
+## Synchronizing the USB access
+
+When a USB data transfer is requested while another operation is already in progress, it will commonly result in a "device is busy" error. Since the preview and the settings UI update regularly, and the user might be trying to capture an image or modify settings at the same time, such conflicts between different operations turned out to be very frequent.
```

Comment:
I'm not sure which one to choose in this particular case. "frequent" is still the best match, because it's almost literally talking about "frequency" (as in, number of "something" per second). I could replace "very" with "highly" if that helps?

Reply:
I'll resolve this one. Try to be careful of using the phrase 'very _something_'. It's overused and as I've said, there are precise verbs available.